cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -143,8 +148,9 @@ private boolean doesLeaderRemoveViolateOptimizedGoal(Replica replicaToBeRemoved)
       // Moving a follower replica does not violate/affect this goal
       return false;
     }
-    int topicLeaderCountOnSourceBroker = replicaToBeRemoved.broker().numLeadersFor(replicaToBeRemoved.topicPartition().topic());
-    return topicLeaderCountOnSourceBroker <= minTopicLeadersPerBroker();
+    final String topic = replicaToBeRemoved.topicPartition().topic();
+    int topicLeaderCountOnSourceBroker = replicaToBeRemoved.broker().numLeadersFor(topic);
+    return _mustHaveTopicMinLeadersPerBroker.containsKey(topic) && topicLeaderCountOnSourceBroker <= minTopicLeadersPerBroker(topic);
```

Comment:
this includes also a bug fix, previously the check 
```return topicLeaderCountOnSourceBroker <= minTopicLeadersPerBroker();``` 
didn't take into account that the goal may not be applicable to the topic

Reply:
I am not sure if this check is needed. This function is used only after `#actionAffectsRelevantTopics(BalancingAction)` is called, which guarantees that the function `#doesLeaderRemoveViolateOptimizedGoal(Replica)` is called only for a replica from an applicable topic.

Maybe we can update the JavaDoc of `#doesLeaderRemoveViolateOptimizedGoal(Replica)` to make it more clear that this function is intended only for a replica from an applicable topic?

Reply:
The issue I encountered (ie tests failing) is when a `INTER_BROKER_REPLICA_SWAP` is tried in https://github.com/linkedin/cruise-control/blob/fa13315ef00889f5ce119cdb2070ca5ec78d3027/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java#L110

`acceptReplicaSwap` calls `doesLeaderRemoveViolateOptimizedGoal` for both `src` and `dst` topics involved in the swap but only one (e.g src) may be defined as applicable for this goal and not the other one (e.g `dst` topic)
https://github.com/linkedin/cruise-control/blob/fa13315ef00889f5ce119cdb2070ca5ec78d3027/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java#L127

In that case `actionAffectsRelevantTopics` returns true correctly as the action involves a topic applicable (src)
 
that's the reason for double checking

Reply:
LE: It's worth mentioning that previously this bug was hidden as ```return topicLeaderCountOnSourceBroker <= minTopicLeadersPerBroker();``` didn't raise an exception but was incorrectly testing the leader count against min.leaders configuration for an incorrect topic

Reply:
Ah got it, good catch! Thanks for the explanation.

Maybe as a final minor improvement, is it possible to skip the computation of `numLeadersFor` if this goal is not applicable to the `topic`?
```
    if (!_mustHaveTopicMinLeadersPerBroker.containsKey(topic)) {
      // Moving a replica from a non-applicable topic does not violate/affect this goal
      return false;
    }
    int topicLeaderCountOnSourceBroker = replicaToBeRemoved.broker().numLeadersFor(topic);
    return topicLeaderCountOnSourceBroker <= minTopicLeadersPerBroker(topic);
```